File: _study/onnote/design_pattern_js_es6/chap06_생성 패턴/1_싱글턴 Singleton/js/operation_singleton.js
CALC.createNameSpace('CALC.singleton.OperationSingleton');

CALC.singleton.OperationSingleton = (function () {
  class OperationSingleton {
    constructor() {
      if (OperationSingleton._instrance) {
        return OperationSingleton._instrance
      }

      OperationSingleton._instrance = this;
    }

    static getInstance() {
      if (!OperationSingleton._instrance) {
        OperationSingleton._instrance = new OperationSingleton();
      }

      return OperationSingleton._instrance;
    }

    operate(operatorType, firstNumber, secondNumber) {
      let answer = 0;
      let operator = null;

      switch (operatorType) {
        case OperationSingleton.ADD_OPERATION:
          answer = firstNumber + secondNumber;
          operator = '+';
          break;
        case OperationSingleton.SUBSTRACT_OPERATION:
          answer = firstNumber - secondNumber;
          operator = '-';
          break;
        case OperationSingleton.MULTIPLY_OPERATION:
          answer = firstNumber * secondNumber;
          operator = '*';
          break;
        case OperationSingleton.DIVIDE_OPERATION:
          answer = firstNumber / secondNumber;
          operator = '/';
          break;
      }

      let result = `${firstNumber} ${operator} ${secondNumber} = ${answer}`;

      this.print(result);
    }

    print(result) {
      console.log(result);
    }

    toString() {
      return 'OperationSingleton';
    }
  }

  OperationSingleton.ADD_OPERATION = 1;
  OperationSingleton.SUBSTRACT_OPERATION = 2;
  OperationSingleton.MULTIPLY_OPERATION = 3;
  OperationSingleton.DIVIDE_OPERATION = 4;

  return OperationSingleton;

}());

File: _study/onnote/design_pattern_py/Part01/chap01_생성 패턴/1_싱글턴 Singleton/operation_singleton.py
class OperationSingleton:

    __instance = None

    ADD_OPERATION = 1
    SUBSTRACT_OPERATION = 2
    MULTIPLY_OPERATION = 3
    DIVIDE_OPERATION = 4

    @classmethod
    def __getInstance(cls):
        return cls.__instance

    @classmethod
    def getInstance(cls, *args, **kargs):
        if not cls.__instance:
            cls.__instance = cls(*args, **kargs)
        return cls.__instance

    def operate(self, operatorType, firstNumber, secondNumber):
        answer = 0
        operator = None

        if operatorType == OperationSingleton.ADD_OPERATION:
            answer = firstNumber + secondNumber
            operator = "+"
        elif operatorType == OperationSingleton.SUBSTRACT_OPERATION:
            answer = firstNumber - secondNumber
            operator = "-"
        elif operatorType == OperationSingleton.MULTIPLY_OPERATION:
            answer = firstNumber * secondNumber
            operator = "*"
        elif operatorType == OperationSingleton.DIVIDE_OPERATION:
            answer = firstNumber / secondNumber
            operator = "/"

        result = str(firstNumber) + " " + operator + " " + str(secondNumber) + " = " + str(answer)

        self.printResult(result)

    def printResult(self, result):
        print(result)
